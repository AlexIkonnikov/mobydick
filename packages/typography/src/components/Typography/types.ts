import {TextProps} from '@npm/mobydick-core';
import {ITextColors} from '@npm/mobydick-styles';
import {ReactChild} from 'react';
import {TextStyle} from 'react-native';

type ITextChild = ReactChild | string | number | false | undefined | null;

type ITextStyle = TextStyle | undefined | false | null;

export type IStyledTextChildren = ITextChild | ITextChild[];

export interface IStyledTextProps extends TextProps {
  children?: IStyledTextChildren;
  font?: TypographyProp;
  style?: ITextStyle | ITextStyle[];
}

export enum TFontSize {
  H1 = 'H1',
  H2 = 'H2',
  H3 = 'H3',
  H4 = 'H4',
  H5 = 'H5',
  XL = 'XL',
  L = 'L',
  M = 'M',
  S = 'S',
  XS = 'XS',
  XXS = 'XXS',
  XXXS = 'XXXS',
}

export type TFont =
  | 'Inter-Regular'
  | 'Inter-Medium'
  | 'Inter-SemiBold'
  | 'Inter-Bold';

export enum TFontWeight {
  Regular = 'Regular',
  Medium = 'Medium',
  SemiBold = 'SemiBold',
  Bold = 'Bold',
}

export const TEXT = 'Text';

export type TFontColorConverter<K> = K extends `${typeof TEXT}${infer R}`
  ? R
  : never;

export type TFontColor = TFontColorConverter<keyof ITextColors>;

export type TSizes = Record<TFontSize, {fontSize: number; lineHeight: number}>;

export type TWeights = Record<TFontWeight, TFont>;

export type TypographyProp = `${TFontWeight}-${TFontColor}-${TFontSize}`;

export type TFontStyle = Pick<
  TextStyle,
  'color' | 'fontSize' | 'fontFamily' | 'lineHeight'
>;
