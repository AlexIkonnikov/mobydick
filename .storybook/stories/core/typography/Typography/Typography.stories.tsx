import React from 'react';
import {Meta, StoryObj} from '@storybook/react-native';

import {
  defaultTextLightColor,
  ITextColors,
  TEXT,
  TFontColor,
  TFontSize,
  TFontWeight,
  Typography as TypographyLib,
} from '@lad-tech/mobydick-core';

const textColorKeys = Object.keys(
  defaultTextLightColor,
) as (keyof ITextColors)[];

const fontColors = textColorKeys.map(item => item.slice(TEXT.length));

export const Basic = ({
  fontWeight,
  fontColor,
  fontSize,
  text,
}: {
  fontWeight: TFontWeight;
  fontColor: TFontColor;
  fontSize: TFontSize;
  text: string;
}) => (
  <TypographyLib font={`${fontWeight}-${fontColor}-${fontSize}`}>
    {text}
  </TypographyLib>
);

const meta: Meta<typeof Basic> = {
  title: 'Core',
  component: Basic,
};

type Story = StoryObj<typeof Basic>;

export const Typography = {
  args: {
    fontWeight: TFontWeight.Regular,
    fontColor: 'Primary',
    fontSize: TFontSize.H1,
    text: 'Ваш текст',
  },
  argTypes: {
    fontWeight: {
      options: Object.values(TFontWeight),
      control: {
        type: 'select',
      },
    },
    fontColor: {
      options: Object.values(fontColors),
      control: {
        type: 'select',
      },
    },
    fontSize: {
      options: Object.values(TFontWeight),
      control: {
        type: 'select',
      },
    },
  },
} satisfies Story;

export default meta;
