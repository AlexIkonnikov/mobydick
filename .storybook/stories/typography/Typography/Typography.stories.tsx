import React from 'react';
import {select, text} from '@storybook/addon-knobs';

import CenterView from '../../CenterView';

import {
  TEXT,
  TFontColor,
  TFontSize,
  TFontWeight,
  Typography,
  defaultTextLightColor,
  ITextColors,
} from '@lad-tech/mobydick-core';

const textColorKeys = Object.keys(
  defaultTextLightColor,
) as (keyof ITextColors)[];

const fontColors = textColorKeys.map(item => item.slice(TEXT.length));

export default {
  title: 'Design System/Typography/Typography',
  decorators: [getStory => <CenterView>{getStory()}</CenterView>],
};

export const Basic = () => (
  <Typography
    font={`${
      select('Начертание', TFontWeight, TFontWeight.Regular) as TFontWeight
    }-${select('Цвет', fontColors as TFontColor[], 'Primary')}-${select(
      'Размер',
      TFontSize,
      TFontSize.H1,
    )}`}>
    {text('Введите что-нибудь', 'Ваш текст')}
  </Typography>
);

Basic.story = {
  name: 'basic',
};
