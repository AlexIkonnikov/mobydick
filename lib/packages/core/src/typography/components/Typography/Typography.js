"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Typography = void 0;
const Text_1 = __importDefault(require("../../../basic/components/Text/Text"));
const hooks_1 = require("../../hooks");
const Typography = ({ children, font = 'Regular-Primary-S', style, ...props }) => {
    const { fontStyle } = (0, hooks_1.useFont)(font);
    return (<Text_1.default style={[fontStyle, style]} {...props}>
      {children}
    </Text_1.default>);
};
exports.Typography = Typography;
