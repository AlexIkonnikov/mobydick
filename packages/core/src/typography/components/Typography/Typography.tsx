import React, {FC} from 'react';

import Text from '../../../basic/components/Text/Text';
import {useFont} from '../../hooks';
import {IStyledTextProps} from '../../types';

export const Typography: FC<IStyledTextProps> = ({
  children,
  font = 'Regular-Primary-S',
  style,
  ...props
}) => {
  const {fontStyle} = useFont(font);

  return (
    <Text style={[fontStyle, style]} {...props}>
      {children}
    </Text>
  );
};
