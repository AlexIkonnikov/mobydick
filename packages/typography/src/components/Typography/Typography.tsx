import React, {FC} from 'react';
import {Text} from '@npm/mobydick-core';

import {useFont} from '../../hooks';
import {IStyledTextProps} from '../../types';

export const Typography: FC<IStyledTextProps> = ({
  children,
  font = 'Regular-Primary-S',
  style,
  ...props
}) => {
  const {fontStyle} = useFont(font);

  return (
    <Text style={[fontStyle, style]} {...props}>
      {children}
    </Text>
  );
};
