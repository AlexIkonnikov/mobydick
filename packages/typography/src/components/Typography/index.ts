export * from './types';
export * from './Typography';
