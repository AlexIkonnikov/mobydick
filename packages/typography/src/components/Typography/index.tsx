import React, {FC, useMemo} from 'react';
import {useTheme} from '@npm/mobydick-styles';
import {Text} from '@npm/mobydick-core';
import {
  ITextColors,
  textColorKeys,
} from '@npm/mobydick-styles/src/colors/types';

import {
  IStyledTextProps,
  TEXT,
  TFontColor,
  TFontSize,
  TFontWeight,
  TSizes,
  TWeights,
} from './types';

const sizes: TSizes = {
  H1: {fontSize: 36, lineHeight: 40},
  H2: {fontSize: 32, lineHeight: 36},
  H3: {fontSize: 28, lineHeight: 32},
  H4: {fontSize: 24, lineHeight: 26},
  H5: {fontSize: 22, lineHeight: 24},

  XL: {fontSize: 20, lineHeight: 24},
  L: {fontSize: 18, lineHeight: 22},
  M: {fontSize: 16, lineHeight: 20},
  S: {fontSize: 15, lineHeight: 19},
  XS: {fontSize: 14, lineHeight: 18},
  XXS: {fontSize: 12, lineHeight: 16},
  XXXS: {fontSize: 10, lineHeight: 14},
};

const weightsToFonts: TWeights = {
  Regular: 'Inter-Regular',
  Medium: 'Inter-Medium',
  SemiBold: 'Inter-SemiBold',
  Bold: 'Inter-Bold',
};

const getSize = (size: TFontSize): {fontSize: number; lineHeight: number} =>
  sizes[size];

const getWeight = (weight: TFontWeight): string => weightsToFonts[weight];

export const fontColors = (textColorKeys as (keyof ITextColors)[]).map(item =>
  item.slice(TEXT.length),
); // нужно для вывода в сторибуке

const Typography: FC<IStyledTextProps> = ({
  children,
  font = 'Regular-Primary-S',
  style,
  ...props
}) => {
  const {currentTheme, colors} = useTheme();

  const fontStyle = useMemo(() => {
    const [weight, color, size] = font.split('-');
    const {fontSize, lineHeight} = getSize(size as TFontSize);
    return {
      color: colors[currentTheme][`${TEXT}${color as TFontColor}`],
      fontFamily: getWeight(weight as TFontWeight),
      fontSize,
      lineHeight,
    };
  }, [font]);

  return (
    <Text style={[fontStyle, style]} {...props}>
      {children}
    </Text>
  );
};

export default Typography;
