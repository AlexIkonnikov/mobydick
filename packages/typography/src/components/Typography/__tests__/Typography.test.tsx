import React from 'react';
import {render} from '@testing-library/react-native';

import {Typography} from '../Typography';

describe('@npm/mobydick-typography/Typography', () => {
  it('renders default correctly', () => {
    const {toJSON} = render(<Typography>My Text</Typography>);
    expect(toJSON()).toMatchSnapshot();
  });
});
