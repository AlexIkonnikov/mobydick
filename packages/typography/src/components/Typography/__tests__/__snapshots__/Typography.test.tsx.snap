// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@npm/mobydick-typography/Typography renders default correctly 1`] = `
<Text
  allowFontScaling={true}
  style={
    Array [
      Object {
        "color": "#20242D",
        "fontFamily": "Inter-Regular",
        "fontSize": 15,
        "lineHeight": 19,
      },
      undefined,
    ]
  }
>
  My Text
</Text>
`;
